libtiledbsoma/src/soma_reader.cc

```diff
@@ -162,6 +185,14 @@ uint64_t SOMAReader::nnz() {
             "[SOMAReader] nnz is only supported for sparse arrays");
     }
 
+    if (timestamp_ &&
+        (timestamp_->first > 0 || timestamp_->second < UINT64_MAX)) {
+        // Use count_cells for timestamped queries for now.
+        // TODO: make optimized method filter fragment(s) on timestamp range
+        // https://github.com/TileDB-Inc/TileDB/blob/6f92d2864440cec17474f2aae22f223147828c6a/tiledb/sm/cpp_api/fragment_info.h#L314
+        return nnz_slow();
+    }
+
```

Comment:
If we enable timestamps by default, we'll need the optimized nnz to filter filter on timestamp range (TODO) or all nnz calls will be slow.

Let me know if you want and help with that.

Reply:
@gspowley OK, agree...does the TODO sound like the correct approach? iterate through the fragment info, filter out those not overlapping the read timestamp range, then proceed with the rest of the current logic using only the remaining fragments.

Reply:
Yes, iterating through the fragment info and filtering on read timestamp range sounds right. We also need to replace the `fragment_info.total_cell_num()` call, because it computes the total number of cells for all loaded fragments. We can manually compute the total_cell_num by calling `fragment_info.cell_num(index)` for each fragment that passed the timestamp filter.

Reply:
@gspowley `FragmentInfo.timestamp_range()` makes it sound like one fragment can contain multiple writes with distinct timestamps. Is that true, or otherwise why would a fragment have a timestamp *range* instead of just a timestamp? My concern is the optimized method is in trouble if the desired read timestamp falls in the middle of a fragment's "timestamp range".

Reply:
@mlin Right, a fragment will have a timestamp range if the fragment was created by a consolidation operation. If no consolidation operations were performed on the array, fragments will have a single timestamp (start and end of the range are the same). I agree, we will need to fall back to a slow nnz if the desired read timestamp falls in the middle of a fragment's timestamp range.